Limited Global Styles: Add E2E test

mmtr opened the pull request:
Part of https://github.com/Automattic/dotcom-forge/issues/1397

#### Proposed Changes

Adds a new E2E test to check that, when Global Styles are limited, the upgrade modal and the upgrade pre-save notice show up in the Site Editor.

Global Styles became a paid feature (available on Premium plans and above) for sites created after Dec 15, 2022. See p7DVsv-fY6-p2.

That means that Free and Personal sites created after that date would have Limited Global Styles. We basically allow them to try Global Styles, but any custom style is not loaded in the frontend.

As part of the Limited Global Styles flow, we show upgrade banners in the Site Editor. However, those are fragile because they rely on a CSS selector to manipulate the DOM and inject those banners, but Gutenberg can change the DOM at any point without further notice.

So with this test, we hope to catch those regressions when new Gutenberg versions are released to WPCOM so we can update the Limited Global Styles implementation with new CSS selectors.

#### Testing Instructions

- Follow setup instructions from https://github.com/Automattic/wp-calypso/tree/trunk/test/e2e
- Run the test:
  - Desktop: `yarn workspace wp-e2e-tests test -- specs/fse/fse__limited-global-styles.ts`
  - Mobile: `yarn workspace wp-e2e-tests test:mobile -- specs/fse/fse__limited-global-styles.ts`

mmtr committed:
Limited Global Styles: Add E2E test

worldomonation reviewed (changes requested):
**question:blocking** - I wanted to clarify the necessity of adding a new E2E test that imposes a decently heavy price on production databases:
- runs on _every commit for every branch_
- creates a new site for every run

As noted in [this post](p6fDka-Ld-p2), E2E tests are very resource and maintenance-heavy. Are we absolutely sure that an E2E is the only way forward to test this feature? If so, must a new site be created in order to test this feature, or can we use an existing site?

Ideally, all questions in the checklist should be answered with a Yes before we proceed with a new E2E test.

mmtr commented:
Thanks for the feedback @worldomonation!

The main reason why the test creates new site is because it checks the Limited Global Styles feature which only affects sites created after Dec 15, 2022 (sites previously created have full access to Global Styles, see p7DVsv-fY6-p2).

However, looks like all the E2E sites have been created before that date, which I discarded using an existing site. Said that, I guess we can probably add a new permanent site to be used by this test, so I'll look into that.

mmtr commented:
> Are we absolutely sure that an E2E is the only way forward to test this feature?

We believe so! Limited Global Styles relies on the DOM rendered by Gutenberg to inject/display the upgrade modal and pre-save notices. We know that it's fragile, because Gutenberg can change the DOM at any point without further notice, breaking the CSS selectors used by our implementation, so hopefully this E2E can catch these regressions.

mmtr committed:
Use existing site

mmtr commented:
Alright, I created a new site for one of the existing E2E users so we no longer need to create a new site during the E2E run.

mmtr committed:
Fix test for mobile viewports

dsas commented on test/e2e/specs/fse/fse__limited-global-styles.ts:
```diff
@@ -0,0 +1,65 @@
+		await fullSiteEditorPage.setGlobalColorStyle( 'Background', {
+			colorName: 'Primary',
+		} );
+		await fullSiteEditorPage.closeSiteStyles();
```
Looking at the artefacts in TC it looks like it's getting stuck here. It's setting the global style but it isn't saving.

mmtr replied:
Thanks! Seems that changing the background color wasn't working always (because the "Primary" color name is not robust enough), and following runs were failing to see the pre-save notice because the previous run didn't restore the original color.

I ended up adding methods to choose a different style variation which should be more robust and updated the flow to use that instead.

mmtr replied:
Seems TC is still stuck here, but I don't understand why. The test is passing locally and I don't see anything obvious in the artifact video that might explain the issue – the save button should be clickable, but somehow it is not clicked. @worldomonation do you know what's going on?

mmtr replied:
Alright, finally was able to figure out that the failure was caused by the cookie banner intercepting the click that hides the "Site updated" notification before styles are saved again.

I'm not sure if the cookie banner should be visible for E2E tests (asked here p1671789652736069-slack-C03N25JPCE4), but I added extra logic to dismiss it if present.

mmtr committed:
Update flow to change style variation instead

mmtr committed:
Select an inactive style variation

mmtr committed:
Split last step

mmtr committed:
Remove pre-save selector

mmtr committed:
Dismiss cookie banner

mmtr committed:
Locate cookie banner within editor frame

mmtr committed:
Remove test from calypso-pr group

mmtr commented:
> runs on _every commit for every branch_

Since the test is only intended to catch regressions before new versions of Gutenberg are merged into WPCOM, I removed it from the `calypso-pr` group, and kept it only on the `gutenberg` group which I think only runs once a day and also as part of the Gutenberg upgrade process.

dsas commented on packages/calypso-e2e/src/lib/components/cookie-banner-component.ts:
```diff
@@ -1,28 +1,36 @@
 	/**
 	 * Constructs an instance of the component.
 	 *
 	 * @param {Page} page The underlying page.
```
this jsdoc needs updating to be Page or Locator too

dsas reviewed (approved):
One nitpick aside, this all lgtm.

worldomonation commented:
> > Are we absolutely sure that an E2E is the only way forward to test this feature?
> 
> We believe so! Limited Global Styles relies on the DOM rendered by Gutenberg to inject/display the upgrade modal and pre-save notices. We know that it's fragile, because Gutenberg can change the DOM at any point without further notice, breaking the CSS selectors used by our implementation, so hopefully this E2E can catch these regressions.

Thanks for clarifying, and sorry for taking my time with the response. I'll review the changes shortly.

mmtr committed:
Fix selectors for Gutenberg 14.8

mmtr committed:
Fix mobile selectors for Gutenberg 14.8

mmtr committed:
Better param name

mmtr committed:
Update jsdoc

mmtr committed:
Update selectors for Gutenberg v14.9.1

mmtr committed:
Add support for a dynamic panel selector depending on the editor context

mmtr commented:
Thanks @worldomonation. I updated the PR to use the new encrypted config file and also changed some selectors to work with the new Gutenberg v14.9.1, so this should be ready for a final review.

worldomonation commented on test/e2e/specs/fse/fse__limited-global-styles.ts:
```diff
@@ -0,0 +1,60 @@
+		// On mobile, site styles is a popover panel that blocks the Save button.
+		// So let's always close site styles first to be safe. :)
+		await fullSiteEditorPage.closeSiteStyles();
+		await fullSiteEditorPage.save( { expectedPreSaveElement: '.wpcom-global-styles-notice' } );
```
**comment:non-blocking**: in my opinion this isn't a preferred approach for a couple of reasons:

1. it leaks the underlying interaction & operation of the framework to the test spec.
2. it is prone to failure when the selectors change.

In my opinion, a better approach would be to transparently handle the pre-save notice within the `FullSiteEdittorPage.save()` method. 

With that said, I am not going to block the merge of this PR on this comment as it is a minor issue.

mmtr replied:
> In my opinion, a better approach would be to transparently handle the pre-save notice within the `FullSiteEdittorPage.save()` method.

The problem with that approach is that the pre-save notice doesn't appears always – it only shows up if a custom style has been set, so I don't know how `FullSiteEdittorPage.save` could handle this automatically since it doesn't know whether the spec has intentionally set a custom style or not.

For instance, the `fse__limited-global-styles.ts` spec has one step that sets a custom style (so we need to confirm that the pre-save notice shows up) and then another one that sets a default style (so we need to confirm that the pre-save notice doesn't show up).

mmtr replied:
Alright, I came up with something to handle the pre-save notice within `FullSiteEdittorPage.save()` in https://github.com/Automattic/wp-calypso/pull/71333/commits/0a8e9c4d207cf83306177101a76de01780b10cd7

The `FullSiteEdittorPage` class knows now whether custom styles are being used or not based on the usage of the `setStyleVariation` method (so the spec no longer needs to explicitly indicate it), which allowed the `save` method to automatically determine whether the pre-save notice should be visible or not.

However, the spec still needs to pass a param to the `save` method to enable this pre-save notice check, since the `save` method is also used by other specs where we cannot deterministically determine if custom styles have been used or not (because they use other methods such as `setGlobalColorStyle` or `setGlobalLayoutStyle` ).

worldomonation commented on test/e2e/specs/editor-tracking/editor-tracking__global-styles-events.ts:
```diff
@@ -164,10 +164,10 @@ describe( DataHelper.createSuiteTitle( 'Editor tracking: Global styles events' )
 			it( 'Update the global background color', async function () {
 				// We can always guarantee a target color event if we click a different one first.
-				await fullSiteEditorPage.setGlobalColorStlye( 'Background', {
+				await fullSiteEditorPage.setGlobalColorStyle( 'Background', {
```
Good catch!

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -63,16 +36,76 @@ const selectors = {
+			desktopPreviewMenuItem: ( target: EditorPreviewOptions ) =>
+				`button[role="menuitem"] span:text("${ target }")`,
+			previewPane: ( target: EditorPreviewOptions ) => `.is-${ target.toLowerCase() }-preview`,
+
+			// Publish
+			publishButton: ( state: 'disabled' | 'enabled' ) => {
+				const buttonState = state === 'disabled' ? 'true' : 'false';
+				return `${ panel } button.editor-post-publish-button__button[aria-disabled="${ buttonState }"]`;
+			},
+
+			// List view
+			listViewButton: `${ panel } button[aria-label="List View"]`,
+
+			// Details popover
+			detailsButton: `${ panel } button[aria-label="Details"]`,
+
+			// Document Actions dropdown
+			documentActionsDropdown: `${ panel } button[aria-label="Show template details"]`,
+			documentActionsDropdownItem: ( itemSelector: string ) => `.popover-slot ${ itemSelector }`,
+			documentActionsDropdownShowAll: `.popover-slot .edit-site-template-details__show-all-button`,
+
+			// Editor settings
+			settingsButton: ( label = settingsButtonLabel ) =>
+				`${ panel } .edit-post-header__settings .interface-pinned-items button[aria-label="${ label }"]`,
+
+			// Undo/Redo
+			undoButton: 'button[aria-disabled=false][aria-label="Undo"]',
+			redoButton: 'button[aria-disabled=false][aria-label="Redo"]',
+
+			// More options
+			moreOptionsButton: ( label = moreOptionsLabel ) =>
+				`${ panel } button[aria-label="${ label }"]`,
+
+			// Site editor save
+			saveSiteEditorButton: `${ panel } button.edit-site-save-button__button`,
+
+			// Nav sidebar
+			navSidebarButton:
+				'button[aria-label="Block editor sidebar"],button[aria-label="Toggle navigation"]',
+		};
```
**question:blocking**: If I'm understanding correctly, this change is due to the changes introduced in Gutenberg 14.9.1 that adds an iframe within the editor field - is that correct?

If so, I would like to take a stab at fixing the issue at the foundational level before we resort to moving selectors inside the class definition, which is something we prescribe against in the [style guide](https://wpcalypso.wordpress.com/devdocs/test/e2e/docs/style_guide.md#no-selectors-within-class-definition).

mmtr replied:
> If I'm understanding correctly, this change is due to the changes introduced in Gutenberg 14.9.1 that adds an iframe within the editor field - is that correct?

No. This is needed because Gutenberg 14.9.1 has introduced a refactored layout for the site editor, which uses an editor toolbar with a DOM that is slightly different to the toolbar in the post editor. However, `EditorToolbarComponent` is used for both editors, so somehow we need to account for those different layouts. 

It's unrelated to the new iframe used for the post editor, since the previous selectors still work there.

mmtr replied:
> the site editor [...] uses an editor toolbar with a DOM that is slightly different to the toolbar in the post editor

To expand on this, the main difference is that the container in the post editor is `.interface-interface-skeleton__header` while the container in the site editor is `.edit-site-layout__header`.

> moving selectors inside the class definition [...] is something we prescribe against in the [style guide](https://wpcalypso.wordpress.com/devdocs/test/e2e/docs/style_guide.md#no-selectors-within-class-definition).

I guess I can keep the selectors outside of the class, and transform all of them to functions that accept the container as parameter:

```
// ...
blockInserterButton: ( panel ) => `${ panel } button[class*="inserter-toggle"]`,
saveDraftButton: ( panel ) => `${ panel } button[aria-label="Save draft"], button[aria-label="Saved"]`,
switchToDraftButton: ( panel ) => `${ panel } button.editor-post-switch-to-draft`,
// ...
```

I honestly think the code is simpler by keeping the selectors within the class, so they can remain as plain strings, but happy to disagree and commit according to the style guide.

mmtr replied:
> I guess I can keep the selectors outside of the class, and transform all of them to functions that accept the container as parameter

https://github.com/Automattic/wp-calypso/pull/71333/commits/4a7b231d94ecb1f975e950ff174d6e3254b7d28b

mmtr committed:
Transparently handle the pre-save notice within the FullSiteEdittorPage.save() method

mmtr commented:
@worldomonation I think I addressed all your feedback. Can you take another look? Thanks!

worldomonation commented:
> @worldomonation I think I addressed all your feedback. Can you take another look? Thanks!

Yes I will take a look today. Sorry for the delay, Gutenberg 14.9.1 has caused issues for specs.

worldomonation commented on packages/calypso-e2e/src/lib/pages/full-site-editor-page.ts:
```diff
@@ -353,10 +374,26 @@ export class FullSiteEditorPage {
+				selectors.limitedGlobalStylesPreSaveNotice
+			);
+			if ( this.hasCustomStyles ) {
+				await limitedGlobalStylesPreSaveNotice.waitFor();
+			} else {
+				const count = await limitedGlobalStylesPreSaveNotice.count();
+				assert.equal( count, 0 );
```
**question:non-blocking** - looks like the use case included so far for `fullSiteEditorPage.save` always has the `checkPreSaveNotices` flag set to true.

Would it make sense to eliminate the root `if` statement to reduce the layers of nesting?

mmtr replied:
That'd break other existing specs using `fullSiteEditorPage.save`. Some examples:
- https://github.com/Automattic/wp-calypso/blob/62bb2d082c6e53fbda048279c40f545330f3d4bf/test/e2e/specs/editor-tracking/editor-tracking__document-actions-events.ts#L148
- https://github.com/Automattic/wp-calypso/blob/62bb2d082c6e53fbda048279c40f545330f3d4bf/test/e2e/specs/editor-tracking/editor-tracking__global-styles-events.ts#L297

worldomonation replied:
Okay, that's fair. Thanks for the heads up on other usages.

worldomonation commented on packages/calypso-e2e/src/lib/components/cookie-banner-component.ts:
```diff
@@ -1,28 +1,36 @@
  */
 export class CookieBannerComponent {
-	private page: Page;
+	private page: Page | Locator;
```
**issue:blocking** - this change would make the `this.page` property not necessarily represent a page. 

Instead of making this a union type, would adding a new class property `this.editorWindow` like what I did in [this PR](https://github.com/Automattic/wp-calypso/pull/72300/files#diff-74efc6b0758ef8840649a0b3686139322157f42f466e196ae18c859c7171070aR31-R34) be something that can be considered?

```
eg.

private page: Page;
private editorWindow: Locator;
```

and in methods within `CookieBannerComponent`, it can be used like:


```
this.editorWindow.locator().click();
```

This way there is no ambiguity as to what `this.page` and `this.editorWindow` refers to.

mmtr replied:
Not sure if adding something so specific to the editor such as `editorWindow` would be a good idea for this cookie banner component which is supposed to work in other contexts than the editor.

The reason why I added this is because there are some circumstances where the cookie banner is not reachable within the page due to an iframe or similar, so we need to provide a more specific container for the cookie to be found.

That means that sometimes, the cookie would be within the `Page` instance and sometimes would be within the `Locator` instance.

In order to unblock this PR, I can add a second property called `container` (so it supports more scenarios than the editor) and then update the method to try locating first the cookie banner within the page and fallback to the given container if not found, but I honestly consider that adds unnecessary complexity.

worldomonation replied:
> Not sure if adding something so specific to the editor such as editorWindow would be a good idea for this cookie banner component which is supposed to work in other contexts than the editor.

AFAIK the CookieBannerComponent as it is implemented ~~only appears on the editor~~ (clarified) seems to be only used to dismiss cookies in the Editor, thus ~~the concern is unnecessary~~ (clarified) to me it does not make sense to worry about needing to use it for non-editor contexts. If we start using the component elsewhere to dismiss cookie banners, then we can cross the bridge at that time.

> I honestly consider that adds unnecessary complexity.

I would argue otherwise - the APIs between these components are designed to be very similar, and keeping the meaning behind terms such as "page" is important. Making the "page" act as a "locator" or "page" that only differs based on the calling method is not keeping this contract.

mmtr replied:
Fair enough! I'll implement these changes as soon as the test site switches back to a block theme.

mmtr replied:
Solved in https://github.com/Automattic/wp-calypso/pull/71333/commits/b43c6dcf7ae7c74411b09c29f29ab454a25785bc

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-site-styles-component.ts:
```diff
@@ -128,6 +133,22 @@ export class EditorSiteStylesComponent {
+	 *
+	 * @param {StyleVariation} styleVariation The style variation to set.
+	 */
+	async setStyleVariation( styleVariation: StyleVariation ): Promise< void > {
+		await this.returnToTopMenu();
+		await this.clickMenuButton( 'Browse styles' );
+		const locator =
+			styleVariation === 'Default'
+				? this.editor.locator( selectors.defaultStyleVariation )
+				: this.editor.locator( selectors.nonDefaultStyleVariation ).first();
```
**comment:non-blocking** - the method name does not accurately portray the actual behaviour - I had expected this to take a parameter `name` that corresponds to the name shown in the sidebar.

mmtr replied:
Unfortunately, the names of the style variations depend on the theme (e.g. the Twenty Twenty-Three theme has Default, Aubergine, Block out, ...; and the Zoologist theme has Default, Black, White, ...), and since we cannot guarantee the site will always have the same theme, we shouldn't hardcode those names (or they'll fail as soon as the site switches to a different theme).

The only thing we know is that all themes has a Default style variation, so we cannot really do anything more than picking the default or any non-default style variation.

worldomonation replied:
> we shouldn't hardcode those names (or they'll fail as soon as the site switches to a different theme).

I see where you're coming from, but I would argue otherwise. 

Explicitly asserting the text of the elements that is being interacted with ensures that absolutely nothing changes in the UI behavior unless the change is expected (eg. renaming styles as part of Gutenberg update). By explicitly stating that we expect the style "Blue" to be present, you can think of the test spec as performing an `expect( condition )`. 

The current implementation is closer to the implicit assertion on the spectrum. The test will surely fail on catastrophic scenarios such as all styles going missing, but it will not fail and alert on less catastrophic but still undesirable scenarios such as some styles going missing. 

We already do a lot of explicit assertions in the manner I describe above throughout the framework, in fact! When we [select themes from the gallery](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/pages/themes-page.ts#L83) or [select a plan by name in PlansPage](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/pages/plans-page.ts#L81).

My suggestion to perform an explicit assertion would have been something like this:

```
await this.returnToTopMenu();
await this.clickMenuButton( 'Browse styles' );
await locator.getByText( name ).click();

// and if you want to assert even more that the style was indeed applied..
await locator.waitFor( 'some sort of selector ');
```

Anyway, since this is a non-blocking comment I will leave it at that and (addendum) respect your choices here.

mmtr replied:
> unless the change is expected

I think this is key part here. For instance, the site used by this PR had its theme recently changed to Twenty Twenty-One which is completely breaking this E2E test since that theme doesn't support the site editor.

That theme switch was a manual action and I don't know if that's what'd we consider an expected change that should affect the E2E tests.

Anyway, do you think we can change the theme back to a block theme such as Twenty Twenty-Three or Zoologist? If the site necessarily needs a legacy theme, we might need to create a a new site.

mmtr replied:
> do you think we can change the theme back to a block theme such as Twenty Twenty-Three or Zoologist? 

I discussed this privately with @worldomonation, and we cannot activate a block theme just yet until https://github.com/Automattic/wp-calypso/pull/72300 is merged.

So I'm going to put this PR on hold for now since we cannot target any site where we can test Limited Global Styles.

mmtr replied:
Solved in https://github.com/Automattic/wp-calypso/pull/71333/commits/cf76064d382a3c3dd68a4959471177a06113a86b

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -3,54 +3,58 @@ import envVariables from '../../env-variables';
 import type { EditorPreviewOptions } from './types';
 
-const panel = 'div.interface-interface-skeleton__header';
+const postEditorHeader = 'div.interface-interface-skeleton__header';
+const siteEditorHeader = 'div.edit-site-layout__header';
```
**suggestion:non-blocking** - FYI, I was able to reliably select the toolbar element using `playwright.getByRole("region")` in the console by launching a debug instance. This will be i18n-safe as I tested this in both Japanese and English UI editors. 

If methods within the `EditorToolbarComponent` is prepended with `this.editor.getByRole("region")` then that should eliminate the need to pass in the `header` to each selector string.

```
eg.
const locator = this.editor.getByRole('region').first().locator('')
```

Up to you if you want to use this suggestion, I won't block merge on this.

mmtr replied:
I think these changes are no longer needed because of https://github.com/Automattic/wp-calypso/pull/72112, so I'll see if I can revert them as soon as we're able to switch the site's theme back to a block theme.

mmtr replied:
I was able to revert these changes thanks to https://github.com/Automattic/wp-calypso/pull/72112.

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -63,16 +67,23 @@ const selectors = {
+	constructor(
+		page: Page,
+		editor: Locator,
+		context: 'post-editor' | 'site-editor' = 'post-editor'
```
**nit:blocking** - the term `context` is quite loaded in Playwright, as it refers to the browsercontext (think of it as a session containing multiple pages which are tabs). 

Would something like `editorType` work instead?

mmtr replied:
I think these changes are no longer needed because of https://github.com/Automattic/wp-calypso/pull/72112, so I'll see if I can revert them as soon as we're able to switch the site's theme back to a block theme.

mmtr replied:
I was able to revert these changes thanks to https://github.com/Automattic/wp-calypso/pull/72112.

worldomonation reviewed (commented):
Apologies for making you wait for the review. I left comments and suggestions where relevant.

Also FYI, there have been changes to the `EditorPage` and its descendant POMs from work to support Gutenberg 14.9.1 in https://github.com/Automattic/wp-calypso/pull/72112. There is also https://github.com/Automattic/wp-calypso/pull/72300 which is pending review and merge that may also cause a conflict.

mmtr commented:
Putting this PR on hold as per https://github.com/Automattic/wp-calypso/pull/71333/files#r1081055836, given that right now we cannot target any site where we can test the Limited Global Styles functionality.

mmtr committed:
Adapt selectors to latest Gutenberg release

mmtr committed:
Add editor property to CookieBannerComponent

mmtr committed:
Use explicit style variations names

worldomonation commented on packages/calypso-e2e/src/lib/pages/full-site-editor-page.ts:
```diff
@@ -34,10 +36,12 @@ const selectors = {
 	templateLoadingSpinner: '[aria-label="Block: Template Part"] .components-spinner',
 	closeStylesWelcomeGuideButton:
 		'[aria-label="Welcome to styles"] button[aria-label="Close dialog"]',
+	limitedGlobalStylesModalTryButton: '.wpcom-global-styles-modal button:has-text("Try it out")',
```
**comment:non-blocking** - this works if the UI is English, but if we ever want to expand FSE testing to non-English then we'd have to adjust the selector. Just a comment, no need to take action.

mmtr replied:
Thanks! I adjusted it in https://github.com/Automattic/wp-calypso/pull/71333/commits/7d92a6b24794b42e8bfee463d697c30774561bf4 so it no longer depends on the English language to be set.

worldomonation commented on packages/calypso-e2e/src/lib/components/cookie-banner-component.ts:
```diff
@@ -1,28 +1,39 @@
+	 * @param {Locator} editor Locator or FrameLocator to the editor.
 	 */
-	constructor( page: Page ) {
+	constructor( page: Page, editor: Locator ) {
```
**comment** - I see what you mean by the cookie notice showing up at a bunch of places in Calypso. So far, it does not seem to interfere with the E2E tests except for this FSE spec, but I'll be keenly watching if it causes flakiness going forward.

This is also just a comment, no need to change or do anything.

mmtr committed:
Make try button selector independent of language

mmtr committed:
Merge remote-tracking branch 'origin/trunk' into add/limited-global-styles-e2e

fullofcaffeine commented:
Hi! 👋🏻 I just noticed this failed for AT sites [here](https://teamcity.a8c.com/buildConfiguration/calypso_WPComTests_gutenberg_atomic_edge_desktop/10210145?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildTestsSection=true). I see this test is hardcoded to a single simple site without any notion of cross-env testing or Gutenberg edge versions. If a feature is related to Gutenberg (by seeing the , then it should _ideally_ be exposed/tested in 4 environments: 

1. simple prod (the simple site you setup for this test here); ✅ 
2. simple edge (missing); ❌ 
3. AT prod (missing); ❌ 
4. AT edge (missing). ❌ 

I'm assuming it is relevant to Gutenberg because it's testing the Site Editor, and it has the `@group gutenberg` metadata comment at the top. You should use the `[getTestAccountByFeature](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts)` API to configure the [criterion](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/utils/criteria-for-test-accounts.ts) for the site instead of hardcoding it, otherwise, it might still be included in unsuitable builds within the same groups. I understand you might not have been aware of this API, and that's okay, but I wanted to raise awareness here. Usually, it's just a matter of setting up the criteria for the site here, and then adding `const accountName = getTestAccountByFeature( envToFeatureKey( envVariables ) )`. This will get the right account name based on the criterion table + env variable passed, that you can then pass to the `TestAccount` constructor. Example: https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/appearance/theme__details-preview-activate.ts#L23.

I can help to add the test site criterion for this test, I'll look into that tomorrow. I'll mute it from the Gutenberg edge builds for now as the current site is not really AT or edge-aware.

mmtr commented:
Thanks @worldomonation. I opened https://github.com/Automattic/wp-calypso/pull/78246 to skip this test for Atomic sites.

> You should use the [getTestAccountByFeature](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts) API

Global Styles are [only limited on sites created after we made made GS a paid feature (2022-12-15, or those with a blog ID higher than 213403000)](https://github.com/Automattic/wp-calypso/blob/126b34c8c030b9f4abcfd622bab7c038d107e4ab/apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/index.php#L364-L367), so the E2E test needs a site created after that date.

If I use `getTestAccountByFeature( { ...envToFeatureKey( envVariables ), variant: 'siteEditor' } )`, I get a production site that was created in 2021-02-05 (blog ID: 188950373), and a edge site created in 2021-08-24 (blog ID: 196891745).

Can we update these tests sites to newer sites where Global Styles are limited so the E2E can use them?
